documentation/examples/custom-sd/adapter-usage/main.go

```diff
@@ -0,0 +1,252 @@
+		}
+		tgroup.Labels = labels
+
+		// Add all key/value pairs from the node's metadata as their own labels.
+		for k, v := range node.NodeMeta {
+			name := strutil.SanitizeLabelName(k)
+			tgroup.Labels[model.LabelName(model.MetaLabelPrefix+name)] = model.LabelValue(v)
+		}
+
+		tgroup.Targets = append(tgroup.Targets, target)
+	}
+	return &tgroup, nil
+}
+
+// Note: you must implement this function for your discovery implementation as part of the
+// Discoverer interface. Here you should query your SD for it's list of known targets, determine
+// which of those targets you care about (for example, which of Consuls known services do you want
+// to scrape for metrics), and then send those targets as a target.TargetGroup to the ch channel.
+func (d *discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
+	for c := time.Tick(time.Duration(d.refreshInterval) * time.Second); ; {
+		var srvs map[string][]string
+		resp, err := http.Get(fmt.Sprintf("http://%s/v1/catalog/services", d.address))
+
+		if err != nil {
+			level.Error(d.logger).Log("msg", "Error getting services list", "err", err)
+			time.Sleep(time.Duration(d.refreshInterval) * time.Second)
+			continue
+		}
+
+		dec := json.NewDecoder(resp.Body)
+		err = dec.Decode(&srvs)
+		resp.Body.Close()
+		if err != nil {
+			level.Error(d.logger).Log("msg", "Error reading services list", "err", err)
+			time.Sleep(time.Duration(d.refreshInterval) * time.Second)
+			continue
+		}
+
+		var tgs []*targetgroup.Group
+		// Note that we treat errors when querying specific consul services as fatal for for this
+		// iteration of the time.Tick loop. It's better to have some stale targets than an incomplete
+		// list of targets simply because there may have been a timeout. If the service is actually
+		// gone as far as consul is concerned, that will be picked up during the next iteration of
+		// the outer loop.
+		for name := range srvs {
+			if name == "consul" {
+				continue
+			}
+			resp, err := http.Get(fmt.Sprintf("http://%s/v1/catalog/service/%s", d.address, name))
+			if err != nil {
+				level.Error(d.logger).Log("msg", "Error getting services nodes", "service", name, "err", err)
+				break
+			}
+			tg, err := d.parseServiceNodes(resp, name)
+			if err != nil {
+				level.Error(d.logger).Log("msg", "Error parsing services nodes", "service", name, "err", err)
+				break
+			}
+			tgs = append(tgs, tg)
+		}
+		if err != nil {
+			// We're returning all Consul services as a single targetgroup.
+			ch <- tgs
+		}
+		// Wait for ticker or exit when ctx is closed.
+		select {
+		case <-c:
+			continue
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func newDiscovery(conf sdConfig) (*discovery, error) {
+	cd := &discovery{
+		address:         conf.Address,
+		refreshInterval: conf.RefreshInterval,
+		tagSeparator:    conf.TagSeparator,
+		logger:          logger,
+	}
+	return cd, nil
+}
+
+func main() {
+	a.Parse(os.Args[1:])
+	a.HelpFlag.Short('h')
+
+	_, err := a.Parse(os.Args[1:])
```

Comment:
you're calling a.Parse 2 times: here and on line 224.

Reply:
you're right, thanks!